### PR TITLE
BUGFIX: Don't re-generate nodeNames on move if not needed.

### PR DESCRIPTION
**Pull request by sbruggmann**
Partially rolling back changes from #2098 because of issues like #2258, neos/neos-development-collection#2284 and neos/neos-development-collection#2282.

### CODE REVIEW

**Comment by dimaip** [comment]
Sorry, I can not provide a review since I don't have a slightest clue about this part of the code. /cc @kitsunet @kdambekalns @bwaidelich and all ya clever CR people :)

File: Classes/Domain/Model/Changes/AbstractMove.php [hunk 1]
```diff
@@ -67,4 +67,20 @@ protected function generateUniqueNodeName(NodeInterface $parentNode)
...
+     *
+     * @param NodeInterface $parentNode
+     * @param NodeInterface $node
+     * @return string
+     */
+    protected function generateUniqueNodeNameIfNeeded(NodeInterface $parentNode, NodeInterface $node)
+    {
+        if ($this->contentRepositoryNodeService->nodePathAvailableForNode($parentNode->getPath() . $node->getName(), $node)) {
+            return $node->getName();
+        }
```
Comment (mficzel) [inline Classes/Domain/Model/Changes/AbstractMove.php]: Should'nt this simple return 'null' if no new name is needed.

Comment (mficzel) [reply Classes/Domain/Model/Changes/AbstractMove.php]: Other than that the PR works for the cases i see.

Comment (sbruggmann) [reply Classes/Domain/Model/Changes/AbstractMove.php]: `null` ends with the same result but is better more transparent :+1:

File: Classes/Domain/Model/Changes/AbstractMove.php [hunk 2]
```diff
@@ -67,4 +67,20 @@ protected function generateUniqueNodeName(NodeInterface $parentNode)
         return $this->contentRepositoryNodeService
             ->generateUniqueNodeName($parentNode->getPath());
     }
+
+    /**
+     * Generate a unique node name for the copied node if needed
+     *
+     * @param NodeInterface $parentNode
+     * @param NodeInterface $node
+     * @return string|null
```
Comment (bwaidelich) [inline Classes/Domain/Model/Changes/AbstractMove.php]: I'm always very skeptical when it comes to mixed return types. It probably makes sense here, but we should document what `null` means exactly

Comment (bwaidelich) [reply Classes/Domain/Model/Changes/AbstractMove.php]: ..and what "if needed" means in this case

**Comment by bwaidelich** [comment]
I left some (not very helpful) comments but just from reading the code it is not clear to me what this actually fixes. Can't we add a test that breaks without the fix?

File: Classes/Domain/Model/Changes/AbstractMove.php [hunk 3]
```diff
@@ -67,4 +67,20 @@ protected function generateUniqueNodeName(NodeInterface $parentNode)
         return $this->contentRepositoryNodeService
             ->generateUniqueNodeName($parentNode->getPath());
     }
+
+    /**
+     * Generate a unique node name for the copied node if needed
```
Comment (kdambekalns) [inline Classes/Domain/Model/Changes/AbstractMove.php]:
```suggestion
     * Generate a (new) unique node name for the copied node if the current name is in use already
```

File: Classes/Domain/Model/Changes/AbstractMove.php (hunk 2 above)
Comment (kdambekalns) [reply Classes/Domain/Model/Changes/AbstractMove.php]:
```suggestion
     * @return string|null If the current name can be used, null is returned
```

File: Classes/Domain/Model/Changes/MoveAfter.php [hunk 4]
```diff
@@ -43,7 +43,7 @@ public function apply()
             $before = self::cloneNodeWithNodeData($this->getSubject());
             $parent = $before->getParent();
 
-            $nodeName = $this->generateUniqueNodeName($this->getSiblingNode()->getParent());
+            $nodeName = $this->generateUniqueNodeNameIfNeeded($this->getSiblingNode()->getParent(), $this->getSubject());
```
Comment (bwaidelich) [inline Classes/Domain/Model/Changes/MoveAfter.php]: What about moving the logic here to make this a little more explicit? So along the lines of:
```
if ($this->requiresNewNodeName(...)) {
  $nodeName = $this->generateUniqueNodeName($this->getSiblingNode()->getParent());
  $this->getSubject()->moveAfter($this->getSiblingNode(), $nodeName);
} else {
  $this->getSubject()->moveAfter($this->getSiblingNode());
}
```

**Comment by kdambekalns** [comment]
Maybe @nlx-lars can test this as well!?

File: Classes/Domain/Model/Changes/MoveAfter.php (hunk 4 above)
Comment (kdambekalns) [reply Classes/Domain/Model/Changes/MoveAfter.php]: I like!

**Comment by bwaidelich** [comment]
Apart from the nitpicks above I wonder if this truely fixes the bug: If you moved a node forth and back and during that a rename _was_ required (because a name was already taken) we'd still end up in an invalid state, right? But maybe there's no easy way around that (except for event sourcing *g)

**Comment by sbruggmann** [comment]
@bwaidelich Correct! **If** two nodeNames collide, or if it's a copy instead of a move, the basic bug still applies.

This is only a fix that the problem in the CR is less critical. But it's still there.

**Comment by nlx-lars** [comment]
> Maybe @nlx-lars can test this as well!?

This does fix the problem for me as well.

**Comment by mficzel** [comment]
@bwaidelich if you trigger a renaming because you had a collision in the node name (13 char randomness) AND move the node back to the old parent node afterward you will still trigger the faulty cr behavior. Currently this is triggered whenever a node is moved to another parent and back before publishing. So this does not fix the cr bug, but will not expose it that much.

**Comment by bwaidelich** [comment]
@sbruggmann @mficzel thanks for clarifying :)

File: Classes/Domain/Model/Changes/MoveAfter.php (hunk 4 above)
Comment (sbruggmann) [reply Classes/Domain/Model/Changes/MoveAfter.php]: Sure I can do that if you wish, I see the benefits of the explicit style.
But this needs to be done three times and then it's just duplicated code.

**Comment by kdambekalns** [comment]
That build failure should be solveable by `yarn update` to get rid of that disappeared dependency. YMMV. :)

**Comment by kdambekalns** [comment]
Ok then…

**Pull request merged**